lib/active_merchant/billing/gateways/barclaycard_smartpay.rb

```diff
@@ -60,7 +60,13 @@ def credit(money, creditcard, options = {})
         post = payment_request(money, options)
         post[:amount] = amount_hash(money, options[:currency])
         post[:card] = credit_card_hash(creditcard)
-
+        if !options.has_key?(:order_id)
```

Comment:
I couldn't find this pattern elsewhere, but it seemed like a really clean way of handling this client-side. Other reasonable options would be to do this, but without the `requires!` line; or just to unconditionally pass through these attributes, if they exist.

Reply:
I've typically seen the pattern of conditionally adding each option individually vs the use of `requires!`. Example:
```
post[:dateOfBirth] = if options[:date_of_birth]
post[:entityType] = if options[:entity_type]
post[:nationality] =  if options[:nationality]
```
I don't have the full scope of changes in my head so I'm assuming that each item can be sent to the gateway individually instead of requiring the entire group of options.

Reply:
> I don't have the full scope of changes in my head so I'm assuming that each item can be sent to the gateway individually instead of requiring the entire group of options.

In this particular case, the absence of an order ID requires _all four options at once_; none are optional, and none are required for other transaction types—hence this style of `requires!`. (More details in the JIRA ticket this branch is named after, if you're bored.) Whether it should be ActiveMerchant's job to do this, and whether shoving a `requires!` halfway through the method, are still points I'm not clear on. It definitely works, but it's a pattern I didn't see in other adapters I glanced at.

Reply:
What's the purpose of the `order_id` check?

Reply:
And yeah it's more common to check the fields individually than with a `requires!` block. We usually let users send whatever they have and leave out what they don't, and get a error message from the gateway. Depends on how the gateway responds to missing or blank fields.

Reply:
The `order_id` check is because those fields aren't required if it's present. The alternative to doing it this way would be to say that it's not ActiveMerchant's problem to tell you you're missing fields, at which point this method would instead become:

```
def credit(money, creditcard, options = {})
  post = payment_request(money, options)
  post[:amount] = amount_hash(money, options[:cgurrency])
  post[:card] = credit_card_hash(creditcard)
  post[:dateOfBirth] = options[:date_of_birth] if options[:date_of_birth]
  post[:entityType]  = options[:entity_type] if options[:entity_type]
  post[:nationality] = options[:nationality] if options[:nationality]
  post[:shopperName] = options[:shopper_name] if options[:shopper_name]
  commit('refundWithData', post)
end
```

This seemed harsher to the user and doesn't save us much work, but it permits greater flexibility on the part of the gateway.